diff -u3 -p -r a/src/event/ngx_event_openssl.c b/src/event/ngx_event_openssl.c
--- a/src/event/ngx_event_openssl.c	2012-11-12 22:00:32.000000000 +0400
+++ b/src/event/ngx_event_openssl.c	2012-12-17 12:13:58.615790994 +0400
@@ -604,6 +604,16 @@ ngx_ssl_handshake(ngx_connection_t *c)
 
     ngx_ssl_clear_error(c->log);
 
+#if (NGX_GRAPHITE)
+    struct timeval tp;
+    if (c->ssl->handshake_process == 0) {
+        c->ssl->handshake_process = 1;
+        ngx_gettimeofday(&tp);
+        c->ssl->handshake_start_sec = tp.tv_sec;
+        c->ssl->handshake_start_msec = tp.tv_usec / 1000;
+    }
+#endif
+
     n = SSL_do_handshake(c->ssl->connection);
 
     ngx_log_debug1(NGX_LOG_DEBUG_EVENT, c->log, 0, "SSL_do_handshake: %d", n);
@@ -667,6 +677,12 @@ ngx_ssl_handshake(ngx_connection_t *c)
 
         c->ssl->handshaked = 1;
 
+#if (NGX_GRAPHITE)
+        c->ssl->handshake_process = 0;
+        ngx_gettimeofday(&tp);
+        c->ssl->handshake_end_sec = tp.tv_sec;
+        c->ssl->handshake_end_msec = tp.tv_usec / 1000;
+#endif
         c->recv = ngx_ssl_recv;
         c->send = ngx_ssl_write;
         c->recv_chain = ngx_ssl_recv_chain;
diff -u3 -p -r a/src/event/ngx_event_openssl.h b/src/event/ngx_event_openssl.h
--- a/src/event/ngx_event_openssl.h	2012-11-13 14:42:16.000000000 +0400
+++ b/src/event/ngx_event_openssl.h	2012-12-14 19:51:39.998667723 +0400
@@ -42,6 +42,14 @@ typedef struct {
     ngx_event_handler_pt        saved_read_handler;
     ngx_event_handler_pt        saved_write_handler;
 
+#if (NGX_GRAPHITE)
+    ngx_uint_t                  handshake_process;
+    time_t                      handshake_start_sec;
+    ngx_msec_t                  handshake_start_msec;
+    time_t                      handshake_end_sec;
+    ngx_msec_t                  handshake_end_msec;
+#endif
+
     unsigned                    handshaked:1;
     unsigned                    renegotiation:1;
     unsigned                    buffer:1;
